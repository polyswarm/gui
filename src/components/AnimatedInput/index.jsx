// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {CSSTransition} from 'react-transition-group';
// Project imports
// Component imports
import strings from './strings';

class AnimatedInput extends Component {
  constructor(props) {
    super(props);
    this.state = {
      focused: false,
      value: '',
      first: true,
    }

    this.onBlur = this.onBlur.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onFocus = this.onFocus.bind(this);
  }

  componentDidMount() {
    let {props: {value}} = this;
    value = value || ''; 
    this.setState({first: false, value: value});
  }
  value
  render() {
    const {state: {focused, value, first}, props: {placeholder, error, input_id, type} } = this;
    const edited_type = type || 'text';
    const altered = this.getValue(first, value);
    const inputClass = classNames('AnimatedInput-Input', {'AnimatedInput-Error': error});
    return (
        <div className='AnimatedInput'>
          <CSSTransition
            in={value || focused}
            timeout={300}
            classNames='label'> 
            {() => (
              <label className='AnimatedInput-Label'
                htmlFor={input_id}>
                {placeholder}
              </label>
            )}
          </CSSTransition>
          <input
            className={inputClass}
            id={input_id}
            type={edited_type}
            value={altered}
            onBlur={this.onBlur}
            onFocus={this.onFocus}
            onChange={this.onChange}/>
          {error && (
            <p className='AnimatedInput-ErrorLabel'>
              {error}
            </p>
          )}
        </div>
    );
  }

  onBlur() {
    this.setState({focused: false});
  }

  onChange(e) {
    const {props: {onChange}} = this;
    const value = e.target.value;
    if (onChange) {
      onChange(value);
    }
  }

  onFocus() {
    this.setState({focused: true});
  }

  /**
   * This triggers a redraw of the element so we get CSSTransition to animate
   * when the nectar is set as a prop for the first draw.
   */
  getValue(first, value) {
    if (first) {
      return '';
    } else {
      return value;
    }
  }
}
AnimatedInput.proptypes = {
  onChange: PropTypes.func,
  value: PropTypes.string,
  error: PropTypes.string,
  type: PropTypes.string,
  input_id: PropTypes.string,
  placeholder: PropTypes.string,
}
export default AnimatedInput;