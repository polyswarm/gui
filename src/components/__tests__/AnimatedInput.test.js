import React from 'react';
import {render, shallow, mount} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import AnimatedInput from '../AnimatedInput';

it('renders without crashing', () => {
  const wrapper = render(<AnimatedInput />);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('calls onchange when instance onChange is called', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<AnimatedInput type='number' onChange={onChange}/>);
  const instance = wrapper.instance();

  instance.onChange({target:{value: 'change'}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith('change');
});

it('calls onchange when value changed', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<AnimatedInput type='number' onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: 'change'}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith('change');
});

it('calls onChange with valid:true when value is over 0 ', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<AnimatedInput type='number' onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: .0000000000000001}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith(.0000000000000001,);
});

it('calls onchange with valid:false when empty', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<AnimatedInput type='number' onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: ''}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith('');
});

it('calls onchange with valid:false value is 0', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<AnimatedInput type='number' onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: 0}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith(0);
});

it('calls onchange with valid:false value is less than 0', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<AnimatedInput type='number' onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: -1}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith(-1);
});

it('adds error class when error is set in props', () => {
  const wrapper = shallow(<AnimatedInput error='error'/>);
  
expect(wrapper.find('.AnimatedInput-Error')).toHaveLength(1);
});

it('shows the error test that is set', () => {
  const wrapper = mount(<AnimatedInput type='number' error='error'/>);
  
  expect(wrapper.find('p').text()).toEqual('error');
});

it('sets the type when put in prop', () => {
  const wrapper = mount(<AnimatedInput type='number'/>);

  expect(wrapper.find('input').props().type).toEqual('number');
});

it('sets the id on the input with type_id', () => {
  const wrapper = mount(<AnimatedInput input_id='i'/>);

  expect(wrapper.find('input').props().id).toEqual('i')
});
