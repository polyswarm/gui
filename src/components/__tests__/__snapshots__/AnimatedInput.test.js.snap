// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="AnimatedInput"
>
  <div>
    
  </div>
  <input
    class="AnimatedInput-Input"
    type="text"
    value=""
  >
    
  </input>
</div>
`;
